Applications/Spire/Source/Ui/TranslatedTableModel.cpp

```diff
@@ -0,0 +1,132 @@
+#include "Spire/Ui/TranslatedTableModel.hpp"
+#include <algorithm>
+
+using namespace boost::signals2;
+using namespace Spire;
+
+TranslatedTableModel::ScopeExit::ScopeExit(std::function<void()> f)
+  : m_f(std::move(f)) {}
+
+TranslatedTableModel::ScopeExit::~ScopeExit() {
+  m_f();
+}
+
+TranslatedTableModel::TranslatedTableModel(std::shared_ptr<TableModel> source)
+    : m_source(source),
+      m_transaction_level(0) {
+  m_index_mapping.resize(source->get_row_size());
+  std::iota(m_index_mapping.begin(), m_index_mapping.end(), 0);
+  m_source_operation_connection = m_source->connect_operation_signal(
+    [=] (const Operation& operation) {
```

Comment:
This does not handle transactions properly. If a transaction is performed on source then its translation must also be a transaction. Add a unit test to handle transactions and then implement a fix for it.

Reply:
There are some unit tests about transaction of source. When a transaction is performed on source, the transaction will be passed into visit function inside the slot and it will be broken down to basic operations.

Reply:
A transaction in the source must be treated as a transaction in the translation.

Reply:
How to handle it? Does the translation deal with all operations in a transaction of the source but not signal them?

Reply:
translaction.connect([&] (const auto& operation) {
      auto transaction = get<Transaction>(&operation);
      REQUIRE(transaction != nullptr);
      REQUIRE(get<AddOperation>(&transaction->m_operations[0]) != nullptr);
      REQUIRE(get<AddOperation>(&transaction->m_operations[1]) != nullptr);
    });
    source.transact([&] {
      source.push({5});
      source.push({6});
    });

Reply:
oh, I see.